src/models.jl

```diff
@@ -0,0 +1,124 @@
+mutable struct Model
+    metadata::ModelMetadata
+    configuration::ModelConfiguration
+    boundary_conditions::BoundaryConditions
+    constants::PlanetaryConstants
+    eos::EquationOfState
+    grid::Grid
+    velocities::VelocityFields
+    tracers::TracerFields
+    pressures::PressureFields
+    G::SourceTerms
+    Gp::SourceTerms
+    forcings::ForcingFields
+    forcing::Forcing
+    stepper_tmp::StepperTemporaryFields
+    operator_tmp::OperatorTemporaryFields
+    ssp  # ::SpectralSolverParameters or ::SpectralSolverParametersGPU
+    clock::Clock
+    output_writers::Array{OutputWriter,1}
+    diagnostics::Array{Diagnostic,1}
+end
+
+"""
+    Model(; kwargs...)
+
+Construct an `Oceananigans.jl` model. The keyword arguments are:
+
+          N (tuple) : model resolution in (x, y, z)
+          L (tuple) : model domain in (x, y, z)
+         dt (float) : time step
+ start_time (float) : start time for the simulation
+  time_step (float) : ?
+         arch (sym) : architecture (:cpu or :gpu)
+  float_type (type) : floating point type for model data (typically Float32 or Float64)
+          constants : planetary constants (?)
+                eos : equation of state to infer density from temperature and salinity
+            forcing : forcing functions for (u, v, w, T, S)
+boundary_conditions : boundary conditions
+     output_writers : output writer
+        diagonstics : diagnostics
+
+        ... and more.
+"""
+function Model(;
+    # Model resolution and domain size
+             N = (32, 1, 32),
+             L = (1.0, 1.0, 1.0),
+    # Molecular parameters
+             ν = 1.05e-6,
+            νh = ν,
+            νv = ν,
+             κ = 1.43e-7,
+            κh = κ,
+            κv = κ,
+    # Time stepping
+            dt = 1e-16,
```

Comment:
I think this should be Δt as it is not an infinitesimal.

Reply:
I am personally fine with that, but worry that unicode will discourage some users. I think we may want to introduce aliases in the APIs. We can do this by writing

```julia
function Model(;
    ...
    dt = 0.1,
    Δt = dt,
   ...
)
```

etc for other Greek symbols. Thereby users who can't type Greek can use the roman alias instead, while users who can type Greek can revel in their pretty scripts unsullied by unnecessary romans.

Reply:
I agree with the `dt` alias because yeah you never know who's using the code. It's 2019!

Maybe it's just me. I will literally google for `Delta` and copy paste it just to have `Δt` in my code. We're raised from kindergarten to think of `dt` as infinitesimal!